Add porting guide (#716)
dotnet/runtime: docs/design/coreclr/botr/guide-for-porting.md

        by category, so this subsetting mechanism will only need to be a
        directory structure system.

    -   Some set of tests will need to be excluded on a test by test basis. Once
        the product is ready to ship, most of these disabled tests will need to
        have been re-enabled, but there are tests which will be disabled for
        months/years as the product is brought up to quality.

    -   Produce crash or core dumps. The failure mode of many tests during this
        phase will be a crash. A test running tool that captures core dumps will
        make these issues easier to diagnose.

    -   Produce bucketized lists of failures. Generally the approach is to group
        by assertion, and if there is a crash, group by callstack of crash.

-   The first test category to focus on is the JIT category, to bring up the
    general ability to run .NET code. Most of these tests are very simple, but
    getting some code to work is a prerequisite for handling more complex
    scenarios. When doing initial bringup, configuring the Gen0 budget of the GC
    to be a large number so that the GC does not attempt to run during most

>     tests is very useful.

Reviewer: (e.g., `COMPlus_GCgen0size=999999999`)

-   Once basic code is executing, the focus shifts to enabling the GC to work.
    In this initial phase, the correct choice is to enable conservative GC

>     tracking via the FEATURE_CONSERVATIVE_GC macro. This feature will make

Reviewer: ```suggestion
    tracking via the `FEATURE_CONSERVATIVE_GC` macro. This feature will make
```

    garbage collection largely function correctly, but it is not suitable for
    production use of .NET, and can under certain circumstances trigger
    unbounded memory use.

-   Once basic GC works, and basic JIT functionality is present, work can fan
    out into all of the various features of the runtime. Of particular interest
    to engineers porting the runtime are the EH, stackwalking, and interop
    portions of the test suite.

-   During this phase, porting the SOS plugin from the
    <https://github.com/dotnet/diagnostics> will be very useful. The various
    commands available via that tool such as dumpmt, dumpdomain and such are
    regularly useful to developers attempting to port the runtime.

Stage 2 Expand scenario coverage
--------------------------------

-   Once the coreclr tests are largely passing, the next step is to enable
    XUnit. At this time the clr is probably mostly capable of running XUnit

>     tests, and adding the tests from the libraries repo will require XUnit to

Reviewer: ```suggestion
    tests, and adding testing using the libraries tests will require XUnit to
```

    work well.

-   Once XUnit is functional, bring up the libraries set of tests. There
    is quite a lot of the CoreCLR codebase that is largely only tested by the
    libraries test suites.

-   Engineers should also begin to attempt real scenario tests at this point,
    such as ASP.NET Core applications. If the libraries test suites work, then
    ASP.NET Core should as well.

Stage 3 Focus on performance
----------------------------

-   Throughput performance at this time is likely to be not that great. There
    are three major opportunities to improve performance at this stage.

    -   Replace conservative GC with precise GC.

>     -   Tune the assembly stubs generate to be high performance on the platform,

Reviewer: ```suggestion
    -   Tune the assembly stubs to be high performance on the platform,
```

        and implement optional assembly stubs where hand-written assembly would
        be faster than the equivalent C++ code.

>     -   Improve the backend code generation of the JIT.

Reviewer: ```suggestion
    -   Improve the code generated by the JIT.
```

    usage on the platform. Implementing the ahead of the time compiler starts to
    be useful at this time to improve startup performance.

Stage 4 Focus on stress
-----------------------

-   Stress testing the system is necessary to provide confidence that the system
    really works.

-   See the various test passes done in CI, but most critically GCStress testing
    is needed. See documentation around use of the ComPlus_GCStress environment
    variable.

Stage 5 productization
----------------------

-   Productization is about making the runtime able to run shipped effectively
    on a platform.

-   This document does not attempt to list out the work here as it is largely
    specific to the platform in use and the opinions of numerous stakeholders.

Design issues
=============

These large architecture specific design issues will have substantial impact on
both the JIT and VM.

1.  Calling convention rules – Caller pop vs Callee pop, HFA arguments,
    structure argument passing rules, etc. CoreCLR is designed to utilize a
    broadly similar ABI to the OS api. Managed to managed calls typically have a

>     small set of tweaks extensions to the ABI for VM efficiency purposes, but it

Reviewer: ```suggestion
    small set of tweaks or extensions to the ABI for VM efficiency purposes, but it
```

    is generally intended that the ABI of managed code and the ABI of native
    code are very similar. (This is not a hard requirement, and on Windows X86
    the runtime supports a managed to managed abi as well as 3 separate native
    abis for interop, but this scheme is generally not recommended.) See the

>     [CLR-ABI](clr-abi.md) document for how the existing architectures work. When 

Reviewer: ```suggestion
    [CLR-ABI](clr-abi.md) document for how the existing architectures work. Ensure that the CLR-ABI document is updated with all the requisite details and special cases of the new platform. When 
```

    defining the behavior of a new processor architecture abi for CoreCLR, we must maintain that:

>     1.  The this pointer is always passed in the same register regardless of

Reviewer: ```suggestion
    1.  The `this` pointer is always passed in the same register regardless of
```

    relocations for use by load, store, jmp and call instructions) See
    <https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#coff-relocations-object-only>
    for the sort of details that need to be defined.

3.  Behavior and accessibility of processor single step features from within a
    process. On Unix the CLR debugger uses an in process thread to single step
    through functions.

4.  Unwind information. CoreCLR uses Windows style unwind data internally, even
    on Unix platforms. A Windows style unwind structure must be defined. In
    addition, it is possible to enable generation of DWARF data for exposure
    through the GDB JIT
    <https://sourceware.org/gdb/onlinedocs/gdb/JIT-Interface.html> . This
    support is conditional on an \#ifdef, but has been used in the past to
    support bring up of new platforms.

5.  EH Funclets. .NET requires a 2 pass exception model in order to properly
    support exception filters. This substantially differs from the typical
    Itanium ABI model which is used on most Linux architectures

6.  OS behavior with Signals. Especially exactly where the reported instruction
    pointer is located.

> 7.  Little vs big endian. While.NET runtimes have been ported to big endian in

Reviewer: ```suggestion
7.  Little vs big endian. While .NET runtimes have been ported to big endian in
```
(add space before ".NET")

    a big endian platform.

Components of the Runtime affected by a port to a new architecture
==================================================================

Notable components

1.  The JIT. The jit maintains the largest concentration of architecture
    specific logic in the stack. This is not surprising. See [Porting RyuJit](porting-ryujit.md)
    for guidance.

2.  The CLR VM. The VM is a mix of completely architecture neutral logic, and
    very machine specific paths.

3.  System.Private.CoreLib/System.Reflection. There is little to no architecture
    specific work here that is necessary for bringup. Nice-to-have work involves
    adding support for the architecture in the
    System.Reflection.ImageFileMachine enum, and the ProcessorArchitecture enum,
    and logic that manipulates it.

4.  PE File format changes to add a new architecture. Also, the C# compiler likely
    also needs a new switch to generate machine specific code for the new 
    architecture.

5.  Crossgen/Crossgen2 - As the AOT compilers that produce machine specific logic
    from general purpose MSIL, these will be needed to improve startup performance.

> 6.  R2RDump - This allows diagnosing issues in pre-compiled code.

Reviewer: coredistools - Necessary for GCStress (if determining instruction boundaries is non-trivial), as well as for SuperPMI asm diffs for JIT development.

CLR VM
------

The VM support for architecture specific logic is encoded in a variety of
different ways.

1.  Entirely architecture specific components. These are held in an architecture
    specific folder.

> 2.  Features which are only enabled on certain architectures. Eg. FEATURE_HFA.

Reviewer: ```suggestion
2.  Features which are only enabled on certain architectures. E.g. `FEATURE_HFA`.
```

> 3.  Adhoc \#if blocks used for specific architectures. As needed these are

Reviewer: ```suggestion
3.  Ad-hoc \#if blocks used for specific architectures. As needed these are
```

    added. The general goal is to keep these to a minimum, but difficulty here
    is primarily driven by what special behavior the processor architecture
    requires.

My recommendation would be to look at how Arm64 is implemented in the VM for the
most up to date model of how to implement a CPU architecture.

> ### Architecture Specific Components

Reviewer 2: There are additional important architecture specific components:
* callingconvention.h / argdestination.h
* regdisp.h
* crosscomp.h
* CONTEXT / KNONVOLATILE_CONTEXT_POINTERS / RUNTIME_FUNCTION structures in pal.h
* Unwinder for managed code (see e.g. unwinder_arm64.cpp/h) 
* gcinfodecoder.cpp/h and gcinfoencoder.cpp/h. GC info format may need to be updated too (IIRC we needed to do that for arm64 to accomodate for the larger number of registers)

Reviewer: Are there also components specific to enabling debugging (and profiling or other diagnostics) that need to be called out?

    instructions emitted varies from stub to stub.

7.  Various stub data structures. Many very simple stubs are not emitted via an
    emission of a stream of bytes, but instead are exceptionally regular, and
    are effectively the same instructions for each different stub, only with
    slightly different data members. Instead of using the StubLinker mechanism,
    the VM instead has structures that represent the entirety of the stub and
    its associate data, and fill in the assembly instructions and data fields
    with a normal constructor call setting magic numbers. In addition to being
    executable, these stubs are often parsed to determine exactly what a given
    function is, what it is doing, where control flow will lead to, etc.

#### virtualcallstubcpu.h

This header is used to provide implementation of various stubs as used by
virtual stub dispatch. These stubs are the lookup, resolver, and dispatch stubs
as described in [Virtual Stub Dispatch](virtual-stub-dispatch.md). These are 
maintained in a separate file from the rest of cgencpu.h for historical reasons,
and for reasons of size (there is quite a lot of logic here.)

> 

Reviewer: Also call out `callingconvention.h`?

> ---

Reviewer 3: Delete this - it adds odd line at the top of the rendered view

Guide for porting .NET to a new processor architecture
---

This document is broken up into 2 major sections.

1.  The various porting stages of porting the .NET Runtime

2.  A technical discussion of the major components affected by a port to a new
    architecture

Porting stages and steps
========================

Porting the .NET Runtime to a new architecture typically follows along the
following path.

As engineering continues along the development path, it is best if the logic can
be placed into the master repository of the runtime as soon as possible. This
will have 2 major effects.

1.  Individual commits are easier to review

2.  Not all approaches for fixing problems will always be considered acceptable.
    It is plausible that a some change may not ever be acceptable to take into

>     the public git repo, and discovering such issues early can avoid large

Reviewer 3: ```suggestion
    the upstream git repo, and discovering such issues early can avoid large
```

Anybody can have public fork with anything they wish. We do not have control over that. upstream is what matters.

    amounts of sunk cost

3.  When some change is made which breaks other platforms, it will be relatively
    simple to identify the break. If changes are held until after all changes
    are complete and the product is fully functional, this work is likely to be

>     much more difficult

Reviewer 3: ```suggestion
    much more difficult.
```

(Multiple places, not commenting on all of them.)

    In this initial phase, the correct choice is to enable conservative GC
    tracking via the FEATURE_CONSERVATIVE_GC macro. This feature will make
    garbage collection largely function correctly, but it is not suitable for
    production use of .NET, and can under certain circumstances trigger
    unbounded memory use.

-   Once basic GC works, and basic JIT functionality is present, work can fan
    out into all of the various features of the runtime. Of particular interest
    to engineers porting the runtime are the EH, stackwalking, and interop
    portions of the test suite.

-   During this phase, porting the SOS plugin from the
    <https://github.com/dotnet/diagnostics> will be very useful. The various
    commands available via that tool such as dumpmt, dumpdomain and such are
    regularly useful to developers attempting to port the runtime.

Stage 2 Expand scenario coverage
--------------------------------

-   Once the coreclr tests are largely passing, the next step is to enable
    XUnit. At this time the clr is probably mostly capable of running XUnit
    tests, and adding the tests from the libraries repo will require XUnit to
    work well.

> -   Once XUnit is functional, bring up the libraries/corefx set of tests. There

Reviewer 3: We try to retire corefx name. This should just say libraries.

    on Unix platforms. A Windows style unwind structure must be defined. In
    addition, it is possible to enable generation of DWARF data for exposure
    through the GDB JIT
    <https://sourceware.org/gdb/onlinedocs/gdb/JIT-Interface.html> . This
    support is conditional on an \#ifdef, but has been used in the past to
    support bring up of

5.  EH Funclets. .NET requires a 2 pass exception model in order to properly
    support exception filters. This substantially differs from the typical
    Itanium ABI model which is used on most Linux architectures

6.  OS behavior with Signals. Especially exactly where the reported instruction
    pointer is located.

7.  Little vs big endian. While.NET runtimes have been ported to big endian in
    the past (notable examples include Mono support various game consoles, and
    POWER, and XNA support on Xbox360) there are no current ports of CoreCLR to
    a big endian platform.

Components of the Runtime affected by a port to a new architecture
==================================================================

Notable components

1.  The JIT. The jit maintains the largest concentration of architecture

>     specific logic in the stack. This is not surprising. See the RyuJit porting

Reviewer 3: Add actual link here? https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/porting-ryujit.md

    is primarily driven by what special behavior the processor architecture
    requires.

My recommendation would be to look at how Arm64 is implemented in the VM for the
most up to date model of how to implement a CPU architecture.

### Architecture Specific Components

There are a variety of architecture specific components that all architectures
must implement.

1.  Assembly Stubs

2.  Cgencpu.h (CPU specific header defining stubs and miscellaneous other CPU
    specific details.)

3.  VSD call stub generation (virtualcallstubcpu.hpp and associated logic)

4.  Precode/Prestub/Jumpstub generation

#### Assembly Stubs

There are many reasons for which the runtime requires various assembly stubs.
Here is an annotated list of the stubs implemented for Unix on Arm64.

1.  Only Performance. Some stubs have alternative implementations in C++ code
    which are used if there isn’t an assembly stub. As compilers have gotten
    better, it has become more reasonable to just use the C++ versions. Often
    the biggest performance cost/win is due to fast paths being written that do
    not require setting up a stack frame. Most of the casting helpers fall in
    this category.

>     1.  JIT_Stelem_Ref – very slightly faster version of

Reviewer 3: ```suggestion
    1.  `JIT_Stelem_Ref` – very slightly faster version of
```

This is the usual formatting used for code identifiers.

        implement the logic described in [Virtual Stub Dispatch](virtual-stub-dispatch.md)

    14. ProfileEnter/ ProfileeLeave/ ProfileTailcall – Used to call function
        entry/exit profile functions acquired through the ICorProfiler
        interface. Used in VERY rare circumstances. It is reasonable to wait to
        implement these until the final stages of productization. Most profilers
        do not use this functionality.

    15. JIT_PInvokeBegin/End – Leave/enter the managed runtime state. Necessary
        for ReadyToRun pre-compiled pinvoke calls, so that they do not cause GC
        starvation

    16. VarargPInvokeStub/ GenericPInvokeCalliHelper Used to support calli
        pinvokes. It is expected that C\# 8.0 will increase use of this feature.
        Today use of this feature on Unix requires hand-written IL. On Windows
        this feature is commonly used by C++/CLI

3.  EH Correctness. Some helpers are written in assembly to provide well known
    locations for NullReferenceExceptions to be generated out of a SIGSEGV
    signal.

    1.  JIT_MemSet, and JIT_MemCpy have this requirement

> #### Cgencpu.h

Reviewer 3: ```suggestion
#### cgencpu.h
```

    There is a small set of apis that are called from general purpose vm code
    (EmitComputedInstantiatingMethodStub, EmitShuffleThunkshared) across
    multiple architectures, and then there are the individual assembly
    instruction emission functions which are architecture specific. The
    StubLinker is used to generate complex stubs, where the set of assembly
    instructions emitted varies from stub to stub.

7.  Various stub data structures. Many very simple stubs are not emitted via an
    emission of a stream of bytes, but instead are exceptionally regular, and
    are effectively the same instructions for each different stub, only with
    slightly different data members. Instead of using the StubLinker mechanism,
    the VM instead has structures that represent the entirety of the stub and
    its associate data, and fill in the assembly instructions and data fields
    with a normal constructor call setting magic numbers. In addition to being
    executable, these stubs are often parsed to determine exactly what a given
    function is, what it is doing, where control flow will lead to, etc.

> #### Virtualcallstubcpu.h

Reviewer 3: ```suggestion
#### virtualcallstubcpu.h
```

This header is used to provide implementation of various stubs as used by
virtual stub dispatch. These stubs are the lookup, resolver, and dispatch stubs
as described in [Virtual Stub Dispatch](virtual-stub-dispatch.md). These are 
maintained in a separate file from the rest of cgencpu.h for historical reasons,
and for reasons of size (there is quite a lot of logic here.)

System.Private.CoreLib
----------------------

### Initial Bring up

In System.Private.CoreLib there is no work necessary for initial bring up.

### Complete support

Complete support involves changing the publicly visible api surface of the
product. Doing so is a process handled via public issues on GitHub and
discussions with the api review board.

-   Adding support for the architecture to the

>     System.Reflection.ImageFileMachine enum, and

Reviewer 3: This is not actually necessary (unless you are on Windows and need to support managed C++).

Author: Managed C++ doesn't actually use this either. It's really just folks who want to look at assembly names and such. Its … fairly marginal in use.

Reviewer 3: I would just say that this optional, or just completely drop this note completely. We have not added one `ProcessorArchitecture.Arm64` even though the Arm64 port is done. It is not worth it to add because of it is not required to make anything work in .NET Core. It would be just a busy work.
